### PR TITLE
Add SignalHandler to Presto

### PR DESCRIPTION
**Pull request by gugalnikov**
- This PR adds a SignalHandler class (based on sun.misc.Signal) which then is instantiated by PrestoServer in order to trap & handle "SIGTERM".

- Related to:
#1234
#1224
[https://prestosql.slack.com/archives/CFSDUNJ4W/p1564690238001800](https://prestosql.slack.com/archives/CFSDUNJ4W/p1564690238001800)

- The main purpose of this functionality is to be used with dockerized Presto, by signaling containers automatically to do graceful shutdown (useful when doing auto-scaling, canary, blue-green, etc.)

- The SignalHandler leverages the already existing GracefulShutdown functionality

- Only workers can be gracefully shutdown by this method

- The SignalHandler class is generic and can be instantiated to handle any other signal

### CODE REVIEW

**Comment by martint** [comment]
I have some concerns with this patch:
* `sun.misc.Signal` may not be available in newer versions of the JDK (we're in the progress of moving to Java 11
* Even if we initiate a graceful shutdown via `GracefulShutdownHandler`, when the VM receives a SIGTERM, it will invoke any registered shutdown hooks. Any "badly" behaved code (e.g., HDFS client) will start shutting down at that point.

**Comment by gugalnikov** [comment]
@martint I've tested with JDK 11 (jdk-11.0.6) and didn't have any trouble; seems sun.misc.Signal will still be available for the foreseeable future, or at least they won't remove it until there's a viable alternative for handling signals in Java.

**Comment by gugalnikov** [comment]
@martint regarding your 2nd observation, I made some changes so the SignalHandler is configurable, so it will only trap the indicated signal if a property named: presto.graceful-shutdown-signal is present; this way we can use for example USR1 (SIGUSR1) instead of SIGTERM

**Comment by martint** [comment]
If this is meant to work with signals other than SIGTERM, what's the purpose of doing it via a signal, as opposed to the supported mechanism of submitting a request to the shutdown URL? If the caller can pick and chose how to issue the kill command, it could as easily do "curl -X PUT ...".

**Comment by gugalnikov** [comment]
@martint as i described above this solution is not really meant for small-scale deployments where of course the API call would be the easiest alternative; we're rather thinking of distributed architectures with large (& automatically scheduled) workloads, where the ability to handle signals is quite beneficial in terms of reducing overhead and configuration needed to perform auto-scaling, canary deployments, blue-green, etc, etc.; in many cases i think SIGTERM would be the signal of choice to handle, but this is also configurable for most workload schedulers such as k8s, Nomad, etc., so it makes sense to have this ability as well on Presto side (e.g. https://learn.hashicorp.com/nomad/update-strategies/handling-signals)

File: presto-main/src/main/java/io/prestosql/server/SignalHandler.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import io.airlift.log.Logger;
+import sun.misc.Signal;
+
+import static java.util.Objects.requireNonNull;
+
+public class SignalHandler
+            implements sun.misc.SignalHandler
+{
+    private static final Logger log = Logger.get(GracefulShutdownHandler.class);
```
Comment (csunwold) [inline presto-main/src/main/java/io/prestosql/server/SignalHandler.java:24]:
```suggestion
    private static final Logger log = Logger.get(SignalHandler.class);
```

File: presto-main/src/main/java/io/prestosql/server/SignalHandler.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        if (config.getGracefulShutdownSignal() != null) {
+            Signal signal = new Signal(config.getGracefulShutdownSignal());
+            Signal.handle(signal, new SignalHandler(shutdownHandler));
+        }
+    }
+
+    public void handle(Signal signal)
+    {
+        log.info(signal.toString().trim() + " raised, requesting graceful shutdown...");
```
Comment (csunwold) [inline presto-main/src/main/java/io/prestosql/server/SignalHandler.java:42]:
```suggestion
        log.info("%s raised, requesting graceful shutdown...", signal.toString().trim());
```

**Comment by colebow** [comment]
:wave: @gugalnikov - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**